esphome/components/adc/adc_sensor.cpp

```diff
@@ -18,7 +18,16 @@ static const char *const TAG = "adc";
 // 13 bits for S3 / 12 bit for all other esp32 variants
 // create a const to avoid the repated cast to enum
 #ifdef USE_ESP32
+// TODO: simply use ADC_WIDTH_BIT_DEFAULT when available in newer IDF
 static const adc_bits_width_t ADC_WIDTH_MAX_SOC_BITS = static_cast<adc_bits_width_t>(ADC_WIDTH_MAX - 1);
+#if ADC_WIDTH_MAX_SOC_BITS == ADC_WIDTH_12Bit
+static const int CUSTOM_SOC_ADC_MAX_BITWIDTH = 12;
+#else
+static const int CUSTOM_SOC_ADC_MAX_BITWIDTH = 13;
+#endif
+// TODO: simply use SOC_ADC_MAX_BITWIDTH when available in newer IDF
+static const int ADC_MAX = pow(2, CUSTOM_SOC_ADC_MAX_BITWIDTH) - 1;
```

Comment:
```suggestion
static const int ADC_MAX = (1 << CUSTOM_SOC_ADC_MAX_BITWIDTH) - 1;
```

Not sure how compilers handle this, but as far as i see the standard doesn't require `pow` to be constexpr, and so this should be safer